### PR TITLE
Non-bubbling event should run non-capture listeners

### LINKED ISSUE #742: Composed and non bubbling event semantic dispatched in shadow tree
What is the expected behavior for a *composed* and *non-bubbling* event when dispatched inside a shadow tree?

When an event listener is attached to the host element, and a composed and non-bubbling event is dispatched within a shadow tree, the listener attached on the host element is invoked. Here is an example: https://jsfiddle.net/pmdartus/6ft9gkc4/7/

This behavior is consistent on all the browsers implementing shadow DOM.

As far as I understand the recent changes made to the dispatch event algorithm (https://github.com/whatwg/dom/pull/686), the behavior changed and now the event will only invoke the host element event handler if the event *bubbles* (step 14). Is it the new expected behavior, or did I missed something?

----

Regardless of the outcome of this question, I think it worth adding more details to the non-normative section about `composed`.

### PR DESCRIPTION
**Pull request by annevk**
This fixes a regression from #686.

Tests: various tests were failing because of this. https://github.com/web-platform-tests/wpt/pull/16307 aligns the remaining failing test with the new model.

Fixes #742.

### CODE REVIEW

**Comment by domenic** [comment]
@pmdartus and/or @Zirro if you have time to review this it would be appreciated.

File: dom.bs [hunk 1]
```diff
@@ -1405,15 +1405,21 @@ for discussion).
     </ol>
 
    <li>
-    <p>If <var>event</var>'s {{Event/bubbles}} attribute is true, then <a for=list>for each</a>
-    <var>struct</var> in <var>event</var>'s <a for=Event>path</a>:
+    <p><a for=list>For each</a> <var>struct</var> in <var>event</var>'s <a for=Event>path</a>:
 
     <ol>
      <li><p>If <var>struct</var>'s <a for=Event/path>target</a> is non-null, then set
```
Comment (rniwa) [inline dom.bs]: I actually don't get this check. How can target ever be null?
Don't we need to check that target is equal to targetOverride instead?
In general, steps 5.9.7-9 seem to be in the need of a bit more clarification as to what they're trying to do.

Comment (annevk) [reply dom.bs]: An event's path consists of a list of structs. target is only non-null for the actual target and any shadow tree hosts that the actual target might be "nested" in.

Comment (rniwa) [reply dom.bs]: I understand that's what this sentence assumes but I don't see how such a condition is met given what's in https://dom.spec.whatwg.org/#concept-event-dispatch

Comment (annevk) [reply dom.bs]: Yeah, but that argument is what becomes "item" on the struct, not "target". See https://dom.spec.whatwg.org/#concept-event-path-append. I agree this is somewhat confusing and I'd appreciate if someone could come up with better variable names as I'm still somewhat stuck.

Comment (rniwa) [reply dom.bs]: Oh, I see. That is indeed very confusing. Maybe we can use different variable names or something? Having item, target, & targetOverride is confusing enough.

**Comment by rniwa** [comment]
Okay, with that clarification, the change looks good. That's basically [what I implemented in WebKit](https://trac.webkit.org/browser/webkit/trunk/Source/WebCore/dom/EventDispatcher.cpp?rev=236002#L76).

**Comment by annevk** [comment]
I had forgotten that I already got some suggestions in #645. I'll see about fixing that after this.

**Comment by rniwa** [comment]
We can probably merge this fix as is though. We make editorial improvements in #645.

**Comment by rniwa** [comment]
FWIW, I added a [test](https://svn.webkit.org/repository/webkit/trunk/LayoutTests/fast/shadow-dom/capturing-and-bubbling-event-listeners-across-shadow-trees.html) in https://trac.webkit.org/browser/webkit/trunk/LayoutTests/fast/shadow-dom/capturing-and-bubbling-event-listeners-across-shadow-trees.html?rev=236002.

Feel free to create a PR to merge it into WPT. I can make one too if you'd like.

**Comment by annevk** [comment]
@rniwa if you could turn that into a PR that'd be great, thanks!

**Comment by rniwa** [comment]
@annevk : Ok, posted a PR on https://github.com/web-platform-tests/wpt/pull/16358

**Pull request merged**